core/src/main/java/hudson/model/AbstractItem.java

```diff
@@ -511,14 +514,48 @@ public boolean hasPermission(Permission p) {
      */
     public synchronized void save() throws IOException {
         if(BulkChange.contains(this))   return;
-        getConfigFile().write(this);
+        synchronized (saving) {
+            saving.add(this);
+        }
+        try {
+            getConfigFile().write(this);
+        } finally {
+            synchronized (saving) {
+                saving.remove(this);
+            }
+        }
         SaveableListener.fireOnChange(this, getConfigFile());
     }
 
     public final XmlFile getConfigFile() {
         return Items.getConfigFile(this);
     }
 
+    private static final Set<AbstractItem> saving = new HashSet<>();
+
+    private Object writeReplace() {
+        synchronized (saving) {
+            if (saving.contains(this)) {
+                return this;
+            } else {
```

Comment:
Could limit the monitor scope to the `contains` call, but should not matter—neither the logging nor the `Replacer.<init>` should block.